ddtrace/internal/coverage/code.py

```diff
@@ -0,0 +1,341 @@
+from types import CodeType
+from types import ModuleType
+import typing as t
+
+from ddtrace.internal.compat import Path
+from ddtrace.internal.coverage._native import replace_in_tuple
+from ddtrace.internal.coverage.instrumentation import instrument_all_lines
+from ddtrace.internal.module import BaseModuleWatchdog
+from ddtrace.vendor.contextvars import ContextVar
+
+
+CWD = Path.cwd()
+
+_original_exec = exec
+
+ctx_covered = ContextVar("ctx_covered", default=None)
+ctx_coverage_enabed = ContextVar("ctx_coverage_enabled", default=False)
+
+
+def collapse_ranges(numbers):
+    # This function turns an ordered list of numbers into a list of ranges.
+    # For example, [1, 2, 3, 5, 6, 7, 9] becomes [(1, 3), (5, 7), (9, 9)]
+    # WARNING: Written by Copilot
+    if not numbers:
+        return ""
+    ranges = []
+    start = end = numbers[0]
+    for number in numbers[1:]:
+        if number == end + 1:
+            end = number
+        else:
+            ranges.append((start, end))
+            start = end = number
+
+    ranges.append((start, end))
+
+    return ranges
+
+
+def collect_code_objects(code: CodeType) -> t.Iterator[t.Tuple[CodeType, t.Optional[CodeType]]]:
+    # Topological sorting
+    q = deque([code])
+    g = {}
+    p = {}
+    leaves: t.Deque[CodeType] = deque()
+
+    # Build the graph and the parent map
+    while q:
+        c = q.popleft()
+        new_codes = g[c] = {_ for _ in c.co_consts if isinstance(_, CodeType)}
+        if not new_codes:
+            leaves.append(c)
+            continue
+        for new_code in new_codes:
+            p[new_code] = c
+        q.extend(new_codes)
+
+    # Yield the code objects in topological order
+    while leaves:
+        c = leaves.popleft()
+        parent = p.get(c)
+        yield c, parent
+        if parent is not None:
+            children = g[parent]
+            children.remove(c)
+            if not children:
+                leaves.append(parent)
+
+
+class ModuleCodeCollector(BaseModuleWatchdog):
+    _instance: t.Optional["ModuleCodeCollector"] = None
+
+    def __init__(self):
+        super().__init__()
+        self.seen = set()
+        self.coverage_enabled = False
+        self.lines = defaultdict(set)
+        self.covered = defaultdict(set)
+
+        # Replace the built-in exec function with our own in the pytest globals
+        try:
+            import _pytest.assertion.rewrite as par
+
+            par.exec = self._exec
+        except ImportError:
+            pass
+
+    def hook(self, arg):
+        path, line = arg
+        if self.coverage_enabled:
+            lines = self.covered[path]
+            if line not in lines:
+                # This line has already been covered
+                lines.add(line)
+
+        if ctx_coverage_enabed.get():
+            ctx_lines = ctx_covered.get()[path]
+            if line not in ctx_lines:
+                ctx_lines.add(line)
+
+    @classmethod
+    def report(cls, ignore_nocover: bool = False):
```

Comment:
There is a fair bit of logic here now. Perhaps moving this into a `report` submodule might make this look a bit more tidy?

Reply:
I give you `report.py` and even throw in `util.py` for free.